core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -388,7 +388,7 @@ class Partition(val topic: String,
     val leaderHWIncremented = inWriteLock(leaderIsrUpdateLock) {
       leaderReplicaIfLocal match {
         case Some(leaderReplica) =>
-          val outOfSyncReplicas = getOutOfSyncReplicas(leaderReplica, replicaMaxLagTimeMs)
+          val outOfSyncReplicas = getOutOfSyncReplicas(leaderReplica, replicaMaxLagTimeMs).filter(r => !r.ignoreIsrCheck)
```

Comment:
@junrao @edoardocomar Sorry for the delayed response and this bug. I think I know the problem now.

I don't think there is any problem with initializing `_lastCaughtUpTimeMs` to 0L in Replica. When a topic is created, controller will send `LeaderAndIsrRequest` with isr=all_replicas. When leader receives this `LeaderAndIsrRequest`, it will call `Partition.makeLeader()`, which does the following:

```
(assignedReplicas - leaderReplica).foreach { replica =>
  val lastCaughtUpTimeMs = if (inSyncReplicas.contains(replica)) curTimeMs else 0L
  replica.resetLastCaughtUpTime(curLeaderLogEndOffset, curTimeMs, lastCaughtUpTimeMs)
}
```

Because all replicas are in isr set, there last caughtup time will be initialized to current time. Up to now we are doing the right things. Because controller determines isr set when a broker is newly elected leader, and a replica's `_lastCaughtUpTimeMs` should be initialized to current time if it is in sync. Otherwise we set `_lastCaughtUpTimeMs` to 0.

The problem is that we does `assignedReplicas.filter(_.brokerId != localBrokerId).foreach(_.updateLogReadResult(LogReadResult.UnknownLogReadResult))`, which sets `lastCaughtUpTimeMs` of all replicas to -1. I missed this when I implemented the patch earlier.

I think the right fix is to do the following in `updateLogResult(...)`. I verified that the test created in this patch can succeed after this fix.

```
    if (logReadResult.info.fetchOffsetMetadata.messageOffset >= logReadResult.leaderLogEndOffset)
      _lastCaughtUpTimeMs = math.max(_lastCaughtUpTimeMs, logReadResult.fetchTimeMs)
    else if (logReadResult.info.fetchOffsetMetadata.messageOffset >= lastFetchLeaderLogEndOffset)
      _lastCaughtUpTimeMs = math.max(_lastCaughtUpTimeMs, lastFetchTimeMs)
```